### PR TITLE
fix: SIGNAL-7090 UPDATE async worker id setting per partition

### PR DESCRIPTION
**Pull request by seungjinstord**
## Related Ticket(s)

SIGNAL-7090

## Checklist

- [x] Code conforms to the [Elixir Styleguide](https://github.com/christopheradams/elixir_style_guide)

## Problem

While trying to use the BroadwayAdapter, I found that something in the AsyncAdapter was pushing the messages to just one partition.

## Details

I was able to trace it back to the `get_or_create_worker` function. I found out that unless the `:id` is customized, and just `AsyncWorker` is used, then it doesn't matter whichever distinct name is used for Registry - the same worker pid is going be to tried to be retrieved. This means even though a different partition is found from AsyncAdapter, when it comes to get the associated worker pid - the first one created will be returned for any subsequent worker retrieval attempt.

Downstream effect is the consuming phase gets slowed down - only one partition will have all of the messages for the consumer group. Meaning, no matter how many consumers you have in the consumer group - only one consumer will be handling all of the messages pushed into the topic, across all of the partitions.

The fix is using a custom id that appends the partition number to the atom `AsyncWorker`. I string-concatenated it and didn't explicitly change it to atom, but I suspect it would be done under the hood.

As to why this was not detected in production, is probably because of the robustness of BEAM - a self-healing of crashed supervisor tree would end up picking random workers to be created, resulting in distributing messages across partitions.

But because for firehose we're passing a lot more messages, this became more visible.

TL;DR - I think this would increase overall performance of production consumption of Kafka messages for anywhere using AsyncAdapter. Meaning, this would increase consumption performance of services that consume from topic `wms-service`, which are GAS and WMS Bridge.

### CODE REVIEW

File: lib/kafee/producer/async_adapter.ex [hunk 1]
```diff
@@ -281,7 +281,10 @@ defmodule Kafee.Producer.AsyncAdapter do
       |> Keyword.put(:topic, topic)
       |> Keyword.put(:partition, partition)
 
-    with {:error, {:already_started, pid}} <- Supervisor.start_child(producer, {AsyncWorker, worker_options}) do
+    custom_child_id = "#{AsyncWorker}#{partition}"
```
Comment (btkostner) [inline lib/kafee/producer/async_adapter.ex:284]: Can you please change this to something like `"#{brod_client(producer)}-AsyncWorker-#{topic}-#{partition}"` so the id is unique across all code. Makes it a bit easier to debug if needed.

File: test/kafee/producer/async_adapter_test.exs [hunk 2]
```diff
@@ -87,5 +88,55 @@ defmodule Kafee.Producer.AsyncAdapterTest do
...
+          assert is_pid(pid)
+          pid
+        end
+
+      assert 5 == worker_pids |> Enum.uniq() |> length()
+
+      # Send more messages
+      messages = BrodApi.generate_producer_message_list(topic, 5)
+      # change partitions
+      messages = messages |> Enum.with_index(1) |> Enum.map(fn {message, idx} -> %{message | partition: idx} end)
```
Comment (btkostner) [inline test/kafee/producer/async_adapter_test.exs:118]: Might be worth making a helper function to wrap this up. Something like `BrodApi.generate_producer_partitioned_message_list(topic, number, partitions // 1)`

File: test/kafee/producer/async_adapter_test.exs [hunk 3]
```diff
@@ -87,5 +88,55 @@ defmodule Kafee.Producer.AsyncAdapterTest do
...
+
+      assert :ok = MyProducer.produce(messages)
+      assert_called(AsyncWorker.queue(_pid, _messages), 10)
+
+      # Assert no new worker processes got created
+      worker_pids_2 =
+        for x <- 1..5 do
+          {:via, Registry, {registry_name, registry_key}} = AsyncWorker.process_name(MyProducer.BrodClient, topic, x)
+          assert [{pid, _value}] = Registry.lookup(registry_name, registry_key)
```
Comment (btkostner) [inline test/kafee/producer/async_adapter_test.exs:127]: I've noticed we've called this a couple of times. Do you think it's worth adding this as a `AsyncWorker.process_pid/3` function?

**Review by btkostner — CHANGES_REQUESTED** [review]
Very nice find!

File: lib/kafee/producer/async_adapter.ex (hunk 1 above)
Comment (seungjinstord) [reply lib/kafee/producer/async_adapter.ex:284]: Good point, will do!

Comment (seungjinstord) [reply lib/kafee/producer/async_adapter.ex]: Done in 79c9e59

File: test/kafee/producer/async_adapter_test.exs (hunk 2 above)
Comment (seungjinstord) [reply test/kafee/producer/async_adapter_test.exs]: Done in [7180121](https://github.com/stordco/kafee/pull/97/commits/71801219a82e2e867d88a0d3ff4a17c684e4f072)

File: test/kafee/producer/async_adapter_test.exs (hunk 3 above)
Comment (seungjinstord) [reply test/kafee/producer/async_adapter_test.exs]: Done in a2683bb

**Review by btkostner — APPROVED** [review]
